<h2>Differences betweenPython 2 and Python 3</h2>
<h3>print() is a function</h3>
<p>Probably most visible diffenrence between Python 2 and Python 3. Command <em>print</em> has been replaced by <em>function</em> print()</p>
<p><strong>Python 2</strong>
</p>
<div class="highlight"><pre><code>print 'Hello, World!'
print('Hello, World!')
print "text", ; print 'print more text on the same line'</code></pre></div><div class="highlight"><pre><code>Hello, World!
Hello, World!
text print more text on the same line</code></pre></div><p><strong>Python 3</strong></p>
<div class="highlight"><pre><code>print('Hello, World!')

print("some text,", end="")
print(' print more text on the same line')</code></pre></div><div class="highlight"><pre><code>Hello, World!
some text, print more text on the same line</code></pre></div><div class="highlight"><pre><code>print 'Hello, World!'</code></pre></div><div class="highlight"><pre><code>File "&lt;ipython-input-3-139a7c5835bd&gt;", line 1
    print 'Hello, World!'
                        ^
SyntaxError: invalid syntax</code></pre></div><h3>Division by integers</h3>
<p>Integer division behaves differently in Python 2 and Python 3.</p>
<p><strong>Python 2</strong></p>
<div class="highlight"><pre><code>print '3 / 2 =', 3 / 2
print '3 // 2 =', 3 // 2
print '3 / 2.0 =', 3 / 2.0
print '3 // 2.0 =', 3 // 2.0</code></pre></div><div class="highlight"><pre><code>3 / 2 = 1
3 // 2 = 1
3 / 2.0 = 1.5
3 // 2.0 = 1.0</code></pre></div><p><strong>Python 3</strong></p>
<div class="highlight"><pre><code>print('3 / 2 =', 3 / 2)
print('3 // 2 =', 3 // 2)
print('3 / 2.0 =', 3 / 2.0)
print('3 // 2.0 =', 3 // 2.0)</code></pre></div><div class="highlight"><pre><code>3 / 2 = 1.5
3 // 2 = 1
3 / 2.0 = 1.5
3 // 2.0 = 1.0</code></pre></div><h3>Unicode Support</h3>
<p>Python 2 has separate types for str (), these are ASCII characters and then unicode (). In Python 3, all str () strings are Unicode characters by default. Next, Python 3 introduces byte support.</p>
<h3>Raise an exception</h3>
<p>In Python 2, there are two ways to raise an exception. Python 3 allows only one.</p>
<p><strong>Python 2</strong></p>
<div class="highlight"><pre><code>raise IOError, "file error"</code></pre></div><div class="highlight"><pre><code>raise IOError, "file error"
Traceback (most recent call last):
  File "&lt;stdin&gt;", line 1, in &lt;module&gt;
IOError: file error</code></pre></div><div class="highlight"><pre><code>raise IOError("file error")</code></pre></div><div class="highlight"><pre><code>raise IOError("file error")
Traceback (most recent call last):
  File "&lt;stdin&gt;", line 1, in &lt;module&gt;
IOError: file error</code></pre></div><p><strong>Python 3</strong></p>
<p>The following notation in Python 3 does not work:</p>
<div class="highlight"><pre><code>raise IOError, "file error"</code></pre></div><div class="highlight"><pre><code>raise IOError, "file error"
  File "&lt;stdin&gt;", line 1
    raise IOError, "file error"
                 ^
SyntaxError: invalid syntax</code></pre></div><p>The only correct entry is this:</p>
<div class="highlight"><pre><code>raise IOError("file error")</code></pre></div><div class="highlight"><pre><code>raise IOError("file error")
Traceback (most recent call last):
  File "&lt;stdin&gt;", line 1, in &lt;module&gt;
OSError: file error</code></pre></div><h3>Exception handling</h3>
<p>The treatment of exceptions has also changed. See how exceptions are handled in Python 2 and Python 3.</p>
<p><strong>Python 2</strong></p>
<div class="highlight"><pre><code>try:
    let_us_cause_a_NameError
except NameError, err:
    print err, '--&gt; our error message'</code></pre></div><div class="highlight"><pre><code>name 'let_us_cause_a_NameError' is not defined --&gt; our error message</code></pre></div><p><strong>Python 3</strong></p>
<div class="highlight"><pre><code>try:
    let_us_cause_a_NameError
except NameError as err:
    print(err, '--&gt; our error message')</code></pre></div><div class="highlight"><pre><code>name 'let_us_cause_a_NameError' is not defined --&gt; our error message</code></pre></div><h3>User inputs</h3>
<p>In Python 2, there are two ways to get user input. input() function and raw_input() function. The problem with the input() function is that it can be potentially dangerous because it does not always return a string.</p>
<p><strong>Python 2</strong></p>
<div class="highlight"><pre><code>Python 2.7.6
[GCC 4.0.1 (Apple Inc. build 5493)] on darwin
Type "help", "copyright", "credits" or "license" for more information.

&gt;&gt;&gt; my_input = input('enter a number: ')

enter a number: 123

&gt;&gt;&gt; type(my_input)
&lt;type 'int'&gt;

&gt;&gt;&gt; my_input = raw_input('enter a number: ')

enter a number: 123

&gt;&gt;&gt; type(my_input)
&lt;type 'str'&gt;</code></pre></div><p><strong>Python 3</strong></p>
<div class="highlight"><pre><code>Python 3.7.0 (default, Aug 24 2018, 20:34:01)
[Clang 9.0.0 (clang-900.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
&gt;&gt;&gt; my_input = input('enter a number: ')
enter a number: 123
&gt;&gt;&gt; type(my_input)
&lt;class 'str'&gt;</code></pre></div>